### PR TITLE
Draft of working observe method

### PR DESCRIPTION
**Pull request by fonnesbeck**
`model.observe` now conditions on passed keyword argument data values; force `log_prob` to use keyword arguments.

### CODE REVIEW

File: pymc4/_model.py [hunk 1]
```diff
@@ -83,8 +83,14 @@ def _evaluate(self):
...
         """Return the log probability of the model."""
 
-        def log_prob(*args):
-            context = contexts.InferenceContext(args, expected_vars=self._forward_context.vars)
+        def log_prob(**kwargs):
+            vars = self._forward_context.vars
+            if set(self._observations.keys()).intersection(set(kwargs.keys())):
+                raise ValueError('double specified var')
+            kwargs.update(self._observations)
+            if len(kwargs) != len(vars):
```
Comment (canyon289) [inline pymc4/_model.py]: Out of scope for this PR but adding note. Having `pm.RandomVariable` objects referenced by key in a dictionary would help with checks in the `log_prob` function

https://github.com/pymc-devs/pymc4/issues/112

File: pymc4/_model.py [hunk 2]
```diff
@@ -83,8 +83,14 @@ def _evaluate(self):
...
         """Return the log probability of the model."""
 
-        def log_prob(*args):
-            context = contexts.InferenceContext(args, expected_vars=self._forward_context.vars)
+        def log_prob(**kwargs):
+            vars = self._forward_context.vars
+            if set(self._observations.keys()).intersection(set(kwargs.keys())):
+                raise ValueError('Passed variable already specified.')
+            kwargs.update(self._observations)
+            if len(kwargs) != len(vars):
```
Comment (brianwa84) [inline pymc4/_model.py]: just realized we should probably check more than just len, otherwise len({'x','y'}) == len({'z','a'})

Comment (fonnesbeck) [reply pymc4/_model.py]: Yes, related to Ravin's comment above. If we moved from a list to a dictionary, we could compare keys.

Comment (fonnesbeck) [reply pymc4/_model.py]: Fixed in #118

**Pull request merged**